### ISSUE TITLE
Create an issue

### ISSUE THREAD

**Issue by phr34k** [dart-lang/sdk#54685]
I would like to bring up an issue with the rather limiting testability infrastructure. Presently there are mock and fake and they do an amazing job in most common scenarios. However for just testing a superclass logic I believe there are shortcomings. In specific there seems no way have create mocks that mix and match stubs and actual implementation details which simplifies testing code for superclasses. My understanding is that constraints are partly driven by lacking/limited support of language and runtime features required.

Consider the following example a rather simplified state machine:

```
class Parser
{
    void uknown() {}
    void data() {}
    void scan( List<int> bytes )  {
        ....
    }
}
```

In csharp I can easilly verify wether the implementation of my parser is correct, because language/runtime infrastructure allow me to mix implementation of superclass together with stubbed methods provided from the mock. 

```
Mock<Parser> mock = new Mock<Parser>(behavior: MockBehavior.Strict);
mock.Setup(s => s.scan(It.IsAny<List<int>>())).CallBase();
mock.Object.scan(new List<int> { 27, 36, 0, 0, 28 });
mock.Verify(s => s.uknown(), Times.Never);
mock.Verify(s => s.data(), Times.AtLeastOnce);
```

In dart the mock and fake libraries have no equivelent. It seems to rely on overriding 'noSuchMethod' to record invocations of unimplemented methods to verify and assert. This works great in scenarios with dependency injection is possible or the status quo but doesn't work outside those scenarios. Language constructs like 'MockParser extends Mock implements Parser' inheritly discard implementation details of the superclass, and replace all of them with the 'noSuchMethod' including the scan method we actually trying to verify.

You have a couple of options:

You can manually create classes for special testing purposes by extending from the super class but that effort would be orthogonal to your testing needs. However is rarely a considered good because this is a factually more verbose and require more effort to maintain to keep in sync with the implementation.

```
class TestingParser extends Parser
{
    void uknown() {  uknownCalled++; }
    void data() { dataCalled++; }
    void scan( List<int> bytes )  {
        ....
    }
}
```

You can refactor the implementation code so you introduce dependecy injection but this actually trades off simplicity in your testing code for complexity in your implementation othogonal to your testing needs.

```
class ParserObserver 
{
    void uknown() {}
    void data() {}
}

class Parser
{
    void scan( List<int> bytes , ParserObserver )  {
        ....
    }
}

class MockParserObsever extends Mock implements ParserObserver  { }

```

None of these options they are particulairly appealing because both scenarios trade of the simplicity versus complexity, ideally you just simply want have a simple implementation and matching simple testing code to verify the behavior. Would it be possible to introduce more runtime/language features improve testing of these type of scenarios?

**Comment by lrhn** [comment]
Is this a short-coming of the mocking library (if so, it should probably be moved to the mockito repository), or something you think should be addressed by the testing framework, or the language?

Would it be sufficient if `Mock` was a mixin, so you could do:
```dart
class MockParserObserver extends Parser with Mock implements ParserObserver {}
```
and that would only mock the methods that were not provided by the `Parser` class?

**Comment by phr34k** [comment]
@lrhn I believe it's a shortcoming of the testing framework and should be resolved by language/compiler teams. I am not the author of any testing libraries myself, but I do understand they use '[noSuchMethod](https://pub.dev/documentation/test/latest/fake/Fake/noSuchMethod.html )'  isn't sufficient to implement the desired behavior. 

Therefore it's also my understanding it would not be sufficient to just change `Mock` to mixin without underlying support from the language/runtime. Consider this scenario:

```
class Test {
    String test() { return "base"; }
    String perform() { return test(); }
}

class Mock {
    List<Invocation> called;
    Object noSuchMethod( Invocation ... ) {  called.add(invocation);  }
}

class MockTest1 extends Test with Mock  { }
```

In the most simplest test, you just want to verify that when you call 'perform()', the method 'test()' is called. In type-theory you do have a type that is subtype of Test, complimented with all implementation details of test and perform inherited, with additional complimented with implementation of Mock. When you call, `perform` - the type has implementation for `perform` & `test` so the intended behavior of Mock is bypassed.

The way I see it, is that problem partly stems from the semantics of 'noSuchMethod' which aren't usefull in this scenarios given it's always the _last_ choice in method selection and requires ommision of the type.  For testing frameworks you'd actually want the ability to intercept invocations regardless of this status so they are similair but two opposing sides.

**Comment by lrhn** [comment]
Let me see if I understand this correctly.

You'd want class like `Test` to exist, and then be able to 
* Create an object extending`Test` (so inheriting the implementations)
* And which can also *intercept* invocations. 
    * That can be used to either replace the invocation entirely, like the current mock.
    * Or to just wrap the existing implementation, to record the invocation or otherwise instrument the execution, but still run the original code.

That would allow the code here to call the original `perform`, and recognize the internal call to `this.test` when it happens.
And since both `perform` and `test` are declared on the same class, these approaches won't work:
* Replacing every method (what `Mock` does), since it won't be testing `perform` then.
* Replacing no existing method (what a mixin `Mock` would do, its `noSuchMethod` only triggers on missing methods), since that won't be able to intercept the call to `test`.
* Wrapping the object (`class TestWrapper { final Test _inner; String test() { _intercept(#test); return _inner.test(); } ..,}`, because the `_inner.perform()` would not call the wrapper's `test`.

Is that correct?

If so, we do need something new.

There is no way to *programmatically* create an overriding method in a subclass, you have to actually declare it.
That is, even if we had an `InterceptorMock` mixin, you'd still need to write something in the subclass, probably something like:
```dart
class TestMock extends Test with InterceptorMock {
  test() => this.$interceptMethod(super.test, #test);
  perform() => this.$interceptMethod(super.perform, #test);
}
```
where the first `Function` argument will call the real member, and the remaining arguments represent the current invocation (they'll perhaps be used as arguments to `Invocation.method`).

It's doable, but a little cumbersome. It does seem like something that can be done with code generation, or through macros.

That suggests, to me, that it's a feature that a mock package (likely `mockito`) could consider adding when macros are available (or sooner, since they already have code generation available.).

>  I believe it's a shortcoming of the testing framework and should be resolved by language/compiler teams.

It's not something that directly applies to the `test` package, which is providing the testing framework, and the result-checking, but does not otherwise care what goes in the test bodies.

And it's not something that directly _requires_ new language features, unless it wants to use macros, which are already being designed. I'm not sure what a language feature would look like (possibly a way to declare a class with a non-default `noSuchMethod` which intercepts all members, even those that are already defined, but that still doesn't give you a way to invoke the super-methods.)

All in all, I think this would be best suited as a request on the [`mockito` package](https://pub.dev/packages/mockito), for a way to intercept calls to existing members, and still call the existing implementation.
(Heck, maybe they already have the feature, I'm not that familiar with the package.)

**Comment by phr34k** [comment]
@lrhn that's correct, however it'd would be far easier through extended support by means of a language or runtime feature as opposed to code generation. This is because code generation risks drifting apart from the implementation (manual regeneration on command line), and also is limited in respecting regulair flow of type declaration rules.

With that I mean, if you declare the class final, under normal circumstances type declarations would prohibit you from generating more subtypes i.e. for testing. Language constructs like 'extends Mock implements Test' bypass these requirements in a sense because it's not a subtype anymore but rather 'duck-typed'.

I agree with you that it doesn't require a language feature per-se, but I do think it'd be the best fit. These simple tests I have outlined, are equivelent to an "hello world" program. I can't come up with any reasonable justification that the simplest kind of tests requires more complexity (code generation et all) than an actual complex tests. To me that suggests that the imposed constraints by the language/runtime aren't _that_ well thought out, and have room for improvement.

I understand, but disagree with you about the test package. The test package cares _enough_ to provide `Fake` - which contradicts the neutrality principle. Behavior wise, a Fake a subset of a Mock so heck there could room for consolidation. And assuming the way `noSuchMethod` works is an implementation detail for the testing framework, it kinda of fits in that scope.

That being said, as the old addage says "multiple ways lead to Rome" I believe it would be good to put this up to discussion here to see how dart or the testing framework could faciliate this.

**Comment by lrhn** [comment]
Manual code generation may drift, but macros should not. They'd be generated when the program is compiled.

I don't see us making a way to circumvent `final` for testing only. If it works, it works whether you're testing or not, the language doesn't know.
The one language feature I could see for that would be a "friend library" that is allowed to ignore some permissions (not sealed, maybe `final`/`base`/`interface`), and maybe even share private names.
The only reason to not just write everything in one library, is that you don't want your main library to have the dependencies of the testing library, so allowing a library to be "friends", which may or may not be in the same program, could help with that.
But any guarantee we allow the friend library to break, means it's a guarantee the compiler can't trust when compiling the main library alone.

(I'm surprised that the test package provides `Fake`. It really is just a helper to create classes implementing an interface, without needing to provide implementation for all interface members. Which doesn't feel like it belongs with the rest. It's "a subset of a a Mock", but then, so is `Object`. The `Fake` class extends `Object` with `noSuchMethod(i) => super.noSuchMethod(i);`, nothing more or less. I personally don't think it fits with the rest of the package, it's completely disjoint from the rest, but I guess it can be occasionally useful.)

**Comment by vsmenon** [comment]
Moving this to the language repo as I think the ask seems to bottom out to a new lang feature.
